test/common/http/filter/ext_authz_test.cc

```diff
@@ -0,0 +1,299 @@
+public:
+  HttpExtAuthzFilterTestBase() {}
+
+  FilterConfigSharedPtr config_;
+  Envoy::ExtAuthz::MockClient* client_;
+  std::unique_ptr<Filter> filter_;
+  NiceMock<MockStreamDecoderFilterCallbacks> filter_callbacks_;
+  Envoy::ExtAuthz::RequestCallbacks* request_callbacks_{};
+  TestHeaderMapImpl request_headers_;
+  Buffer::OwnedImpl data_;
+  Stats::IsolatedStoreImpl stats_store_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Network::Address::InstanceConstSharedPtr addr_;
+  NiceMock<Envoy::Network::MockConnection> connection_;
+};
+
+class HttpExtAuthzFilterTest : public testing::Test, public HttpExtAuthzFilterTestBase {
+public:
+  HttpExtAuthzFilterTest() {}
+
+  void SetUpTest(const std::string json) {
+    envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+    MessageUtil::loadFromJson(json, proto_config);
+    config_.reset(new FilterConfig(proto_config, local_info_, stats_store_, runtime_, cm_));
+
+    client_ = new Envoy::ExtAuthz::MockClient();
+    filter_.reset(new Filter(config_, Envoy::ExtAuthz::ClientPtr{client_}));
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+  }
+
+  const std::string filter_config_ = R"EOF(
+    {
+      "grpc_service": {
+          "envoy_grpc": { "cluster_name": "ext_authz_server" }
+      },
+      "failure_mode_allow": true
+    }
+  )EOF";
+};
+
+typedef envoy::config::filter::http::ext_authz::v2::ExtAuthz CreateFilterConfigFunc();
+
+class HttpExtAuthzFilterParamTest : public TestWithParam<CreateFilterConfigFunc*>,
+                                    public HttpExtAuthzFilterTestBase {
+public:
+  virtual void SetUp() {
+    envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config = (*GetParam())();
+    config_.reset(new FilterConfig(proto_config, local_info_, stats_store_, runtime_, cm_));
+
+    client_ = new Envoy::ExtAuthz::MockClient();
+    filter_.reset(new Filter(config_, Envoy::ExtAuthz::ClientPtr{client_}));
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+  }
+};
+
+template <bool failure_mode_allow_value>
+envoy::config::filter::http::ext_authz::v2::ExtAuthz GetFilterConfig() {
+  const std::string json = R"EOF(
+    {
+      "grpc_service": {
+          "envoy_grpc": { "cluster_name": "ext_authz_server" }
+      },
+    }
+  )EOF";
+  envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+  MessageUtil::loadFromJson(json, proto_config);
+  proto_config.set_failure_mode_allow(failure_mode_allow_value);
+  return proto_config;
+}
+
+INSTANTIATE_TEST_CASE_P(ParameterizedFilterConfig, HttpExtAuthzFilterParamTest,
+                        Values(&GetFilterConfig<true>, &GetFilterConfig<false>));
+
+TEST_P(HttpExtAuthzFilterParamTest, NoRoute) {
```

Comment:
style: for each test case/function, please add a short comment explaining what is being tested.

Reply:
Thanks for the review. I've added comments.